scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/pskstore/AdvancedPskStore.java

```diff
@@ -0,0 +1,136 @@
+
+/**
+ * New advanded PSK store with support for asynchrounous PSK store and HSM.
+ * 
+ * Returns psk secret result instead of PSK's secret key. The secret must either
+ * be a master secret (algorithm "MAC"), or a PSK secret key (algorithm "PSK").
+ * If required, the psk secret result maybe returned asynchrounos using a
+ * {@link PskSecretResultHandler}.
+ * 
+ * <p>
+ * Synchronous example returning the PSK secret key:
+ * 
+ * <pre>
+ * &#64;Override
+ * public PskSecretResult generateMasterSecret(ConnectionId cid, ServerNames serverNames, PskPublicInformation identity,
+ * 			String hmacAlgorithm, SecretKey otherSecret, byte[] seed) {
+ * 		SecretKey pskSecret = ... func ... identity ...; // identity maybe normalized!
+ * 		return new PskSecretResult(cid, identity, pskSecret);
+ * }
+ * </pre>
+ *
+ * Asynchronous example returning the master secret:
+ * 
+ * <pre>
+ * &#64;Override
+ * public PskSecretResult generateMasterSecret(ConnectionId cid, ServerNames serverNames, PskPublicInformation identity,
+ * 			String hmacAlgorithm, SecretKey otherSecret, byte[] seed) {
+ * 	
+ * 		start ... func ... cid, servernames, identity, otherSecret, seed 
+ * 			// calls processResult with generate master secret asynchronous;
+ * 		return null; // returns null for asynchronous processing
+ * }
+ * 
+ * &#64;Override
+ * public void setResultHandler(PskSecretResultHandler resultHandler) {
+ * 		this.resultHandler = resultHandler;
+ * }
+ * 
+ * private void processResult(PskPublicInformation identity, ConnectionId cid,
+ * 			SecretKey masterSecret) {
+ * 		PskSecretResult result = new PskSecretResult(cid, identity, masterSecret);
+ * 		resultHandler.apply(result);
+ * }
+ * </pre>
+ * </p>
+ */
+public interface AdvancedPskStore {
+
+	/**
+	 * Check, if HSM supported ECDHE PSK cipher suites.
+	 * 
+	 * @return {@code true}, if ECDHE PSK cipher suites are supported,
+	 *         {@code false}, if not.
+	 */
+	boolean hasEcdhePskSupported();
+
+	/**
+	 * Generate master secret.
+	 * 
+	 * Either return result, or {@code null} and return an asynchronous create
+	 * {@link PskSecretResult} with the CID and master secret calling the result
+	 * handler, provided during {@link DTLSConnector} initialization by
+	 * {@link #setResultHandler(PskSecretResultHandler)}.
+	 * 
+	 * @param cid connection id for stateless asynchronous implementations.
+	 * @param serverName server names. Maybe {@code null}, if SNI is not enabled
+	 *            or not used by the client.
+	 * @param identity psk identity. Maybe normalized
+	 * @param hmacAlgorithm HMAC algorithm name for PRF.
+	 * @param otherSecret other secert from ECDHE, or {@code null}. Must be
+	 *            cloned for asynchrounous use.
+	 * @param seed seed for PRF.
+	 * @return master secret result, or {@code null}, if result is provided
+	 *         asynchronous.
+	 */
+	PskSecretResult generateMasterSecret(ConnectionId cid, ServerNames serverName, PskPublicInformation identity,
```

Comment:
I forgot, to adapt this the the extended function. It may return a master secret or the PSK key.